archive/demos/shiny-drum-machine-audio.js

```diff
@@ -0,0 +1,307 @@
+    return this._data.swingFactor;
+  }
+
+  set tempo(tempo) {
+    this._data.tempo = tempo;
+  }
+
+  get tempo() {
+    return this._data.tempo;
+  }
+
+  getNotes(instrumentName) {
+    const index = 1 + INSTRUMENTS.findIndex((i) => i.name === instrumentName);
+    return this._data[`rhythm${index}`];
+  }
+
+  toggleNote(instrumentName, rhythmIndex) {
+    const notes = this.getNotes(instrumentName);
+    const note = (notes[rhythmIndex] + 1) % 3;
+    notes[rhythmIndex] = note;
+  }
+
+  getNote(instrumentName, rhythmIndex) {
+    const notes = this.getNotes(instrumentName);
+    return notes[rhythmIndex];
+  }
+}
+
+class Player {
+  constructor(beat, onNextBeat) {
+    this.beat = beat;
+    this.onNextBeat = onNextBeat;
+
+    let finalMixNode = context.destination;
+
+    if (context.createDynamicsCompressor) {
+      // Create a dynamics compressor to sweeten the overall mix.
+      const compressor = context.createDynamicsCompressor();
+      compressor.connect(finalMixNode);
+      finalMixNode = compressor;
+    }
+
+    // Create master volume.
+    this.masterGainNode = context.createGain();
+    // Reduce overall volume to avoid clipping.
+    this.masterGainNode.gain.value = 0.7;
+    this.masterGainNode.connect(finalMixNode);
+
+    // Create effect volume.
+    this.effectLevelNode = context.createGain();
+    this.effectLevelNode.gain.value = 1.0; // effect level slider controls this
+    this.effectLevelNode.connect(this.masterGainNode);
+
+    // Create convolver for effect
+    this.convolver = context.createConvolver();
+    this.convolver.connect(this.effectLevelNode);
+  }
+
+  playNote(instrument, rhythmIndex) {
+    this.playNoteAtTime(instrument, rhythmIndex, 0);
+  }
+
+  playNoteAtTime(instrument, rhythmIndex, noteTime) {
+    const note = this.beat.getNote(instrument.name, rhythmIndex);
+
+    if (!note) {
+      return;
+    }
+
+    // Create the note
+    const voice = context.createBufferSource();
+    voice.buffer = this.beat.kit.buffer[instrument.name];
+    voice.playbackRate.value = this.beat.getPlaybackRate(instrument.name);
+
+    let finalNode = voice;
+
+    // Optionally, connect to a panner.
+    if (instrument.pan) {
+      const panner = context.createPanner();
+      // Pan according to sequence position.
+      panner.setPosition(0.5 * rhythmIndex - 4, 0, -1);
+      finalNode.connect(panner);
+      finalNode = panner;
+    }
+
+    // Connect to dry mix
+    const dryGainNode = context.createGain();
+    dryGainNode.gain.value =
+        VOLUMES[note] * instrument.mainGain * this.beat.effect.dryMix;
+    finalNode.connect(dryGainNode);
+    dryGainNode.connect(this.masterGainNode);
+
+    // Connect to wet mix
+    const wetGainNode = context.createGain();
+    wetGainNode.gain.value = instrument.sendGain;
+    finalNode.connect(wetGainNode);
+    wetGainNode.connect(this.convolver);
+
+    voice.start(noteTime);
+  }
+
+  /** Call when beat `n` is played to schedule beat `n+1` */
```

Comment:
Perhaps `/**` to  `//`? Does not look like a JSDoc function notation.

Reply:
Done.